---
title: "生成各级物种分类汇总——STAMP输入文件包括完整分类级"
author: "刘永鑫(Yong-Xin Liu)"
date: '`r Sys.Date()`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
 
## 读取输入文件

```{r read}
# 1. 读取OTU表
otutab = read.table("otutab.txt", header=T, row.names= 1, sep="\t", comment.char = "", stringsAsFactors = F)
# 2. 读取物种注释
tax = read.table("taxonomy.txt", header=T, row.names= 1, sep="\t",comment.char = "", stringsAsFactors = F) 

# 数据按实验设计手动筛选(可选)
metadata = read.table("metadata.txt", header=T, row.names=1, sep="\t") 
# 如筛选soiltypesubspecies列中HIND和HTEJ两组
sub_metadata = subset(metadata, groupID %in% c("LIND","LTEJ","HIND","HTEJ"))
# 实验设计与输入文件交叉筛选
idx = rownames(sub_metadata) %in% colnames(otutab)
sub_metadata = sub_metadata[idx,]
sub_otutab = otutab[,rownames(sub_metadata)]

write.table(sub_metadata,file = "metadata.txt",quote = F,sep = '\t', row.names = T, col.names = T)


# OTU丰度筛选阈值，默认0.1%，0为来筛选
thre = 0.1
# 输出文件名前缀
prefix = "tax"

# 生成各分类级汇总特征表
suppressWarnings(suppressMessages(library(amplicon)))
format2stamp(sub_otutab, tax, thre, prefix)
# 在当前目录生成tax_1-8共7个级别+OTU过滤文件
```

使用此脚本，请引用下文：

If used this script, please cited:

**Yong-Xin Liu**, Yuan Qin, **Tong Chen**, et. al. A practical guide to amplicon and metagenomic analysis of microbiome data. **Protein Cell**, 2021(12) 5:315-330, doi: [10.1007/s13238-020-00724-8](https://doi.org/10.1007/s13238-020-00724-8)

Copyright 2016-2021 Yong-Xin Liu <metagenome@126.com>
